### ISSUE TITLE
Add option to NOT store key on server

### ISSUE THREAD

**Issue by maxhq** [nextcloud/end_to_end_encryption#8]
The current design includes storing the private key encrypted on the server.
Some users with very strong security requirements might prefer to e.g. have a hardware token as the only copy of a key.
If the design would allow an option to not store the client key on the server, hardware token support could be integrated later on.

**Comment by tobiasKaminsky** [comment]
Indeed we want to make this optional. So the user would have to enter a key file on setup of an already initialized account.
On initial setup then also the "publish private key" can be optional.

As this is possible without changing the metadata file, I think this can added without problems later.

We did it first this way as it is risky if you loose your private key and for most user the encrypted and uploaded private key is enough.

@LukasReschke please correct me if I am wrong on any point.

**Comment by schiessle** [comment]
I agree with @tobiasKaminsky. That's on the road map but will be probably a v2.0 feature.

**Comment by sunjam** [comment]
Pardon the silly question, but why is the server app currently needed when the work is being done on the client device?

**Comment by tobiasKaminsky** [comment]
Server app is
- handling encrypted private/public keys, otherwise user would have to transfer them from one device to another manually
- stores/updates encrypted metadata per folder

It is acting like a "stupid" file server, and does not have any real logic.

**Comment by jonathancross** [comment]
Is there work being done to allow use of keys on an a client-side hardware token (eg Yubikey)?

**Comment by sunjam** [comment]
Hardware token sounds like a good new issue to post if nothing comes up in
search.

**Comment by sunjam** [comment]
> handling encrypted private/public keys, otherwise user would have to transfer them from one device to another manually

This is exactly what I would find preferable related to this issue: Only allow users to transfer & recover keys between devices manually (if they prefer it that way).  Thank you all for working on this project!  :+1:

**Comment by ca5ua1** [comment]
As [Joshua Pettus](https://apps.nextcloud.com/apps/end_to_end_encryption) mentioned, if something happens with NC DB, E2EE files would be broken. I believe it's more preferable to give user the trust to keep key secure (since they anyway must keep mnemonic phrase), by making it a default option.

In addition to that, can't we just use mnemonic phrase as a private key (or reproduce keys from it)?

**Comment by susnux** [comment]
> if something happens with NC DB, E2EE files would be broken

Well the threat model of e2ee includes malicious admins.
So even having access to the NC instance will not compromise security of your files.

---

Clients can support this by allowing HSM with pkcs11.
See https://github.com/nextcloud/end_to_end_encryption_rfc/blob/f93dcd87dd57e7fe8c5b5abfd2c1f188bdb4e071/RFC.md#initial-device-with-hardware-usb-token-storage-of-certificate
